pkg/deploy/controller/deploymentconfig/controller.go

```diff
@@ -69,87 +70,89 @@ func (c *DeploymentConfigController) Handle(config *deployapi.DeploymentConfig)
-	latestDeploymentExists, _ := deployutil.LatestDeploymentInfo(config, existingDeployments)
-	if latestDeploymentExists {
-		return nil
-	}
-
-	// check to see if there are inflight deployments
-	if inflightDeployment != nil {
+	// Wait for deployment cancellations. We don't want to compete with
+	// deployment process scaling or lauch competing processes.
+	if inflightDeploymentsExist {
+		c.recorder.Eventf(config, "DeploymentAwaitingCancellation", "Deployment %d awaiting cancellation of running deployments", config.LatestVersion)
 		// raise a transientError so that the deployment config can be re-queued
-		glog.V(4).Infof("Found previous inflight deployment for %s - will requeue", deployutil.LabelForDeploymentConfig(config))
 		return transientError(fmt.Sprintf("found previous inflight deployment for %s - requeuing", deployutil.LabelForDeploymentConfig(config)))
 	}
-
-	// Try and build a deployment for the config.
+	// If the latest deployment already exists, reconcile replica counts.
+	if latestDeploymentExists {
+		// If the latest deployment exists and is still running, there's nothing
+		// to do.
+		if !deployutil.IsTerminatedDeployment(latestDeployment) {
+			return nil
+		}
+		// Reconcile existing deployment replica counts which could have diverged
+		// outside the deployment process.
+		activeDeployment := deployutil.ActiveDeployment(config, existingDeployments)
+		for _, deployment := range existingDeployments.Items {
+			oldReplicaCount := deployment.Spec.Replicas
+			newReplicaCount := oldReplicaCount
+			// The active deployment should be scaled to the config replica count, and
+			// everything else should be at zero.
+			if activeDeployment != nil && deployment.Name == activeDeployment.Name {
+				newReplicaCount = config.Template.ControllerTemplate.Replicas
+			} else {
+				newReplicaCount = 0
+			}
+			if oldReplicaCount != newReplicaCount {
+				deployment.Spec.Replicas = newReplicaCount
+				_, err := c.deploymentClient.updateDeployment(deployment.Namespace, &deployment)
+				if err != nil {
+					c.recorder.Eventf(config, "DeploymentScaleFailed", "Failed to scale %d from %d to %s: %s", deployment.Name, deployment.Spec.Replicas, newReplicaCount, err)
+					return err
+				}
+				c.recorder.Eventf(config, "DeploymentScaled", "Scaled %s from %d to %d", deployment.Name, oldReplicaCount, newReplicaCount)
+			}
+		}
+		return nil
+	}
+	// No deployments are running and the latest deployment doesn't exist, so
+	// create it.
 	deployment, err := c.makeDeployment(config)
 	if err != nil {
 		return fatalError(fmt.Sprintf("couldn't make deployment from (potentially invalid) deployment config %s: %v", deployutil.LabelForDeploymentConfig(config), err))
 	}
-
-	// Compute the desired replicas for the deployment. Use the last completed
-	// deployment's current replica count, or the config template if there is no
-	// prior completed deployment available.
 	desiredReplicas := config.Template.ControllerTemplate.Replicas
```

Comment:
I would like a big fat comment above explaining what has changed since this pretty big.